speechbrain/lobes/models/wav2vec.py

```diff
@@ -0,0 +1,610 @@
+    connecting elements.
+    """
+
+    def __init__(
+        self,
+        embedding_dim,
+        final_dim,
+        latent_extractor=W2VLatentExtractor(),
+        latent_encoder=W2VLatentEncoder(),
+        positional_encoding=PositionalEncoding,
+        target_projector=nn.Linear,
+        feats_projector=nn.Linear,
+        quantiser=GumbelVectorQuantizer(512, 320, (2.0, 0.25, 0.999995,), 2, 256),
+        dropout_context_input=0.0,
+        dropout_targets=0.0,
+        freeze=False,
+    ):
+        """ do_mask should be True just for pretraining! """
+        super().__init__()
+        self.extractor_out_dim = latent_extractor.out_dim
+        assert embedding_dim == latent_encoder.embedding_dim, "Embedding dimension of wav2vec wrapper and latent encoder not equal! You probably modified one component without changing the other."
+        if quantiser is not None:
+            assert quantiser.input_dim == self.extractor_out_dim, "Dimension between quantiser and latent extractor not equal! You probably modified one component without changing the other."
+            assert quantiser.vq_dim == final_dim, "Dimension between quantiser output and model output not equal! You probably modified one component without changing the other."
+
+        self.latent_extractor = latent_extractor
+        self.latent_encoder = latent_encoder
+        self.positional_encoding = positional_encoding(embedding_dim)
+        self.latent_projector = W2VLatentProjector(
+            self.extractor_out_dim, embedding_dim
+        )
+        if quantiser is not None:
+            self.quantiser = quantiser
+            self.target_projector = target_projector(final_dim, final_dim)
+        else:
+            self.target_projector = target_projector(self.extractor_out_dim, final_dim)
+
+        self.feats_projector = feats_projector(embedding_dim, final_dim)
+        self.dropout_context_input = nn.Dropout(dropout_context_input)
+        self.dropout_targets = nn.Dropout(dropout_targets)
+
+        self.mask_emb = nn.Parameter(
+            torch.FloatTensor(embedding_dim).uniform_(), requires_grad=True
+        )
+
+        self.freeze = freeze
+
+    def forward(self, x, x_lens=None, padding_mask=None, mask=None, is_pretraining=False, do_final_projection=False,
+                normalize_signal=True, mask_prob=0., mask_length=10, mask_channel_prob=0.0, mask_channel_length=16,
+                normalize_output=False):
+        """
+        Arguments
+        ----------
+        x : torch.Tensor
+            The raw audio with shape (B,T,).
+        x_lens : torch.Tensor
+            The lengths for each sample of the batch (0<x_lens<1).
+        """
+        if normalize_signal:
+            x = F.layer_norm(x, x.shape[1:])
+
+        results = {}
+        latents = self.latent_extractor(x)
+
+        latent_l2 = latents.float().pow(2.).mean()
+
+        with torch.cuda.amp.autocast(enabled=False):
```

Comment:
Why ?

Reply:
had some issues in the past without, might not be necessary, need to investigate more